File: hyeanwooRoh/programmers/네트워크.js
// 19분

function solution(n, computers) {
  let answer = 0;
  const visit = new Array(n).fill(false);

  function bfs(com) {
    const q = [com];
    visit[com] = true;

    while (q.length !== 0) {
      const tmp = q.shift();
      computers[tmp].forEach((c, idx) => {
        if (c === 1 && !visit[idx]) {
          q.push(idx);
          visit[idx] = true;
        }
      })
    }
  }

  for (let i = 0; i < n; i++) {
    if (!visit[i]) {
      bfs(i);
      answer++;
    }
  }
  
  return answer;
}

console.log(solution(3, [[1, 1, 0], [1, 1, 0], [0, 0, 1]]));
console.log(solution(3, [[1, 1, 0], [1, 1, 1], [0, 1, 1]]));

File: chelseashin/Programmers/08_DFS&BFS/네트워크.py
# 17:00 start
# 17:18 finish
# BFS로 풀이

from collections import deque

def solution(n, computers):
    answer = 0
    visited = [0] * n
    for i in range(n):
        if visited[i]:
            continue
        visited[i] = 1
        Q = deque([i])
        while Q:
            x = Q.popleft()
            for j in range(n):
                if computers[x][j] and not visited[j]:
                    visited[j] = 1
                    Q.append(j)
        answer += 1
    return answer


# DFS로 풀이 - 로직은 BFS와 똑같음

# def dfs(start, computers, visited, n):
#     for i in range(n):
#         if computers[start][i] and not visited[i]:
#             visited[i] = 1
#             dfs(i, computers, visited, n)
#
# def solution(n, computers):
#     answer = 0
#     visited = [0] * n
#     for i in range(n):
#         if visited[i]:
#             continue
#         visited[i] = 1
#         dfs(i, computers, visited, n)
#         answer += 1
#     return answer

print(solution(3, [[1, 1, 0], [1, 1, 0], [0, 0, 1]]))
print(solution(3, [[1, 1, 0], [1, 1, 1], [0, 1, 1]]))